File: TypeScript/DynamicProgramming/memoizedKnapsack.ts
function memoizedKnapsack(
  W: number,
  wt: number[],
  val: number[],
  n: number
): number {

  // dp table initialized with -1
  const dp: number[][] = Array.from({ length: n }, () =>
    Array(W + 1).fill(-1)
  );

  function helper(index: number, remWeight: number): number {
    // base case
    if (index >= n || remWeight === 0) {
      return 0;
    }

    // return cached result
    if (dp[index][remWeight] !== -1) {
      return dp[index][remWeight];
    }

    // recursive case
    const exclude: number = helper(index + 1, remWeight);

    let include: number = 0;
    if (wt[index] <= remWeight) {
      include = val[index] + helper(index + 1, remWeight - wt[index]);
    }

    // store result
    dp[index][remWeight] = Math.max(exclude, include);
    return dp[index][remWeight];
  }

  return helper(0, W);
}

File: Python/DynamicProgramming/memoizedKnapsack.py
def memoizedKnapsack(W, wt, val, n):
    # dp table initialized with -1
    dp = [[-1] * (W + 1) for _ in range(n)]

    def helper(index, rem_weight):
        # base case
        if index >= n or rem_weight == 0:
            return 0

        # return cached result
        if dp[index][rem_weight] != -1:
            return dp[index][rem_weight]

        # recursive case
        exclude = helper(index + 1, rem_weight)
        include = 0

        if wt[index] <= rem_weight:
            include = val[index] + helper(index + 1, rem_weight - wt[index])

        # store result
        dp[index][rem_weight] = max(exclude, include)
        return dp[index][rem_weight]

    return helper(0, W)